packages/hardhat-ledger/src/internal/cache.ts

```diff
@@ -0,0 +1,44 @@
+import type { Paths } from "env-paths";
+import fs from "fs-extra";
+import path from "path";
+
+type CacheableJson = Record<string, string>;
+
+export const CACHE_FILE_NAME = "accounts.json";
+
+export async function write<T extends CacheableJson>(json: T) {
+  const ledgerCacheFile = await getLedgerCacheFile();
+  await fs.writeJSON(ledgerCacheFile, json);
+}
+
+export async function read<T>(): Promise<T | undefined> {
+  const ledgerCacheFile = await getLedgerCacheFile();
+  try {
+    await fs.access(ledgerCacheFile, fs.constants.F_OK);
+    const file: T = await fs.readJSON(ledgerCacheFile);
+    return file;
+  } catch (error) {}
+}
+
+async function getLedgerCacheFile(fileName = CACHE_FILE_NAME): Promise<string> {
+  const ledgerCacheDir = await getLedgerCacheDir();
+  return path.join(ledgerCacheDir, fileName);
+}
+
+async function getLedgerCacheDir(): Promise<string> {
+  const cache = await getCacheDir();
+  const compilersCache = path.join(cache, "ledger");
+  await fs.ensureDir(compilersCache);
+  return compilersCache;
+}
+
+async function getCacheDir(): Promise<string> {
+  const { cache } = await generatePaths();
+  await fs.ensureDir(cache);
+  return cache;
+}
+
+async function generatePaths(packageName = "hardhat"): Promise<Paths> {
```

Comment:
Makes sense! I toyed with the idea of just not taking a param really (as they are internal/private methods) but ended up with this implementation.

I'll change it from [here](https://github.com/NomicFoundation/hardhat/pull/4008/files#diff-b9d3db0dc755826c3c3a8070393addbedd62d6d1497e724804ee9430fa2dff5fR23) too! Let me know what you think.

‼️ This is completely out of scope but I copied the idea from `/hardhat-core/src/internal/util/global-dir.ts` in case you keep track of things like this to change later (or you prefer I change it there too ofc)

Reply:
Ah, I see. Yeah, I'm tempted to change it, but it feels out of scope. Thanks for pointing that out though.